### DISCUSSION TITLE
Recommended Practices to Protect Against Breaking Changes

### DISCUSSION THREAD

**Discussion by RegaCaska** [mlflow/mlflow#12959] [Q&A]
### Wow this is a long post? What's your point?
**How does MLflow envision users manage changes in model signatures breaking clients of said models?**

### Background

I am currently working with a few engineers in my organization to develop an MLOps platform, of which MLFlow is very likely to be used. However, we have a few core issues plaguing our adoption. The one I would like to focus on in this post is protecting our systems against breaking changes. This specifically seems to target model integration (eg. real-time serving) signatures, their enforcement, and how they play into organizing models and versions inside the model registry.

### An Example Problem

Say we iterate on a model, logging to the tracking server with a signature. At the end of it, we have a good baseline model which takes as inputs ("feat_A", "feat_B", "feat_C"). It looks good, so we register the model to the model registry as version 1. A common practice I am seeing is to alias models using "champion" and "challenger". Since this is our baseline model, we will alias it `@champion`.

We continue to iterate on the model, and change our feature selection. We now have a new version of the model logged to the tracking server which takes as inputs ("feat_A", "feat_B", "feat_D") . In other words, we added a new feature "feat_D" and dropped feature "feat_C". Initial tests demonstrate this model may be better than our baseline model, so we register it as version 2 in the model registry and alias it as `@challenger` (We will only support a single challenger for now as that is an separate discussion).

The issue we now encounter is that model versions 1 and 2 have different signatures. This means that a model's signature cannot be used as a system interface contract. Any downstream consumers of our model now have to be updated to support version 2, aka the current `@challenger`. Not the worst at this point, but what happens when we decided that version 2 is actually better than version one, and want to promote it to @champion. This is necessarily a breaking change and puts us in a dilemma. **When we promote to @champion, all consumers break. This means we need a synchronized release with all consumers.**

### Solutions

There are a lot of components which feed into this, but I do think a lot of them all come down to how feature selection is handled. The system interface contract (what client systems pass to use our model, e.g. user_id) and the model signature (which features our model uses such as user_country, user_age, user_salary_bracket, user_aggregate_sentiment etc.) have opposing goals. The system interface needs to be stable while the model signature should be flexible — easy for data scientists to add, remove, or change features freely.

A naïve approach could be to tie model signature to the model registration. This doesn't seem to be the route MLFlow is taking; which is understandable as I can see the registry getting pretty messy after a short while. The next approach would be to tie signature version to the alias, eg. `@v1-champion`, `@v2-champion` etc. This seems to be better, but client systems will have to keep running to stay up-to-date; and we now have an easy point of confusion on differentiating an "official" model version from an "unofficial" naming standard alias/contract version.

What is really needed is an anti-corruption layer between the "raw" model, and the final registered model.

Databricks tries this with their feature_engineering solution. There, they develop a wrapper around your model which knows to reach out either to the online/offline feature store to grab the augmented features. There are a few downsides there. First: you are pretty vendor-locked to databricks, and batch scoring needs to be done (best we can see) via feature_engineering.score_batch. Inherently, the issue there is that the model is heavily tied to the infrastructure; so any infrastructural changes also inherently lead to "retraining" a model.

Really, that exposes a violation of SRP. Mlflow is relatively well suited for "raw" models, while "wrapping" should really be handled via containerization, which just throws us back to square 1.

### Final Words
A final comment I wanted to add on. **Is there documentation on MLFlow's vision for how it works as a cohesive system?** From my foray into the space, there are a lot of tutorials all targeted on a single part of mlflow, but I can't find anything describing how the community envisions its cohesive usage. Put another way: a bag of features rather than a cohesive solution.

I do find external medium articles and blog posts on how people use mlflow, but they all read as hyper-focused on their specific usage of the solution rather than a platform-level view. There doesn't seem to be much consideration on questions like managing model contracts and integrations like discussed here, or composing or chaining models, etc. There seems to be a reliance on heavy investment in long-term dedicated maintenance of models that would get in the way of democratizing their usage.

**Comment by phitoduck** [comment]
I've been pursuing this same set of questions. Did you ever find a satisfying answer?

Love your characterization of the MLflow docs. It's hard not to feel like the "bag of features" approach to the docs is a reflection that MLflow itself does not have a unifying product vision. (I hope that's not true 😐)

For those who don't know: MLflow supports a concept they call "model signatures", which in my own words are metadata (not sure if they're serializable to JSON) which let you add more granular information about the data types that a `model.predict(...)` function accepts and returns. 

E.g. it's not just a `pd.DataFrame`, it's a `pd.DataFrame` with cols `sepal_length` and `sepal_width` and the dtypes of those columns are float.

But to @RegaCaska's point, I can't find a utility of these signatures besides: `model.predict(...)` will raise an exception at inference-time if the inputs don't conform to signature... inference time is a *liiiittle* late to discover that you've introduced a breaking change to your entire system! (specifically: you deployed an API with a breaking contract change, so all clients now break).

Jim Dowling (CEO of Hopsworks) had a [good comment on this](https://www.linkedin.com/posts/jim-dowling-206a98_tough-mlops-problem-ive-never-been-able-activity-7263925520721969153-wbCF?utm_source=share&utm_medium=member_desktop). I'm still understanding, but I think Hopsworks has the concept of a  "Feature Group", which IIUC is a versioned set of columns and transformations to produce them, which can be associated with a model at training time. 

I believe this separates the "system interface" with the "model interface", because if you, e.g. input `user_id` to the system, then the feature store would use that to look up and/or calculate real time the set of features needed by that particular model. FeatureForm might do the same.
